pyproject.toml

```diff
@@ -32,131 +51,131 @@ strict_equality = true
 strict_concatenate = true
 warn_return_any = true
 disallow_subclassing_any = true
-#disallow_untyped_decorators = true
-#disallow_any_generics = true
+disallow_untyped_decorators = true
+disallow_any_generics = true
 
 [[tool.mypy.overrides]]
 module = ["craft-parts"]
-#disallow_untyped_defs = true
+disallow_untyped_defs = true
```

Comment:
This PR is enormous and encompasses a lot of changes (mostly typing changes but I see lots of subtle changes too that I fear are going to get lost in the size of this PR).  

For the ticket size and acceptance criteria, I don't think enabling more mypy rules should be in this PR.  

Perhaps this PR should just do the work in the ticket:

> Acceptance Criteria
Update to the latest ruff that supports all the pylint features that we are currently not checking with ruff